### ISSUE TITLE
passive tree high CPU usage while program idle

### ISSUE THREAD

**Issue by sthalik** [Openarl/PathOfBuilding#618]
The software uses 6% of one core while idle, as long as the passive tree is the current tab.

There are several options here -- use a rendertarget or detect window not visible via `WindowFromPoint`.

My random guess without code inspection is that the individual passive tree elements are drawn each frame. When there are no graphical updates to the tree this would be done needlessly.

If you're not too interested in that I'll try implementing that.

**Comment by Openarl** [comment]
Is 6% of one core high? I wouldn't have thought so. Is that with the program active, or in the background? The program does frequently re-render the entire UI, but the framerate is significantly lowered when in the background, so it shouldn't consume a noticeable amount of CPU time when in the background.

**Comment by sthalik** [comment]
Apologies, it's 6% of a quad-core system, as per Windows' confusing display. This happens even if the program is explicitly maximized.

Actually even 6% of a core would be high. It's the standard among applications to sleep and wait for events, not performing heavy needless work otherwise. 

Furthermore, even when another tab's selected, the application uses 1% of the system's CPU time. It doesn't sleep in the message pump by any means...

Could you point me to the file with the drawing logic? I'm going to send you a pull request anyway, but that could save some of the time.

**Comment by sthalik** [comment]
I see the functionality's inside `SimpleGraphic.dll`. Given you haven't released the source I'm giving up.

**Comment by sthalik** [comment]
I was wrong, there's source code for `SimpleGraphic.dll` in #28.

It's `PeekMessage` in `sys_main.cpp` that's causing the issue. Ideally you'd switch the drawing code to a larger library already handling it, like Qt5 with offscreen opengl widgets.

**Comment by sthalik** [comment]
Please see #636. I have the solution ready.

**Comment by krab** [comment]
Insane CPU usage when program active but mouse cursor outside of window https://i.imgur.com/2HIJxgz.jpg

**Comment by sthalik** [comment]
@krab does this 30% usage also happen with my `SimpleGraphic` fix from #636?

**Comment by krab** [comment]
program don't even work with replaced dll file from your dropbox https://github.com/Openarl/PathOfBuilding/issues/636#issuecomment-332073608

original dll size around 700kb, your dll 10mb?

**Comment by sthalik** [comment]
@krab please the same URL again. I updated it. It had some library dependency problems. Also, there's a new Lua function `RenderInit()` that I had to stub out.

The original DLL was so big since it was a debug build.

**Comment by krab** [comment]
Yep with new DLL program uses 0% cpu when nothing going on in program, up to 33% on my cpu when i move mouse cursor very fast over skill tree. Up to 25% on normal skill tree building/allocating nodes.

**Comment by sthalik** [comment]
The original DLL has an artificial delay when processing events. It makes all input less responsive.

Theoretically we could coalesce events before calling `OnFrame`, but inducing mouse move event all the time just proves a point, I don't see it a use-case to optimize for.

**Comment by jmunson** [comment]
Just wanted to comment to point out that this is still an issue. You can even demonstrate the issue with windows's alt tab window preview. If you have PoB in the background and another window on top, then hold alt and hit tab but leave alt pressed so the alt-tab dialog stays up, then move your mouse around, you can see PoB updating its window in the background as if you were mousing over it, despite it not being the active window.

While most people have the resources to burn, it's still inefficient and when you consider how many third party programs people often need to run while playing PoB, all this inefficiency adds up.
